### PR TITLE
Report improved performance test results to Slack

### PR DESCRIPTION
**Pull request by lptr**
Send a notification to a Slack channel when a performance test shows a significant improvement.

### CODE REVIEW

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/buildcache/TaskOutputCachingJavaPerformanceTest.groovy [hunk 1]
```diff
@@ -117,7 +118,8 @@ class TaskOutputCachingJavaPerformanceTest extends AbstractTaskOutputCachingPerf
         result.assertCurrentVersionHasNotRegressed()
 
         where:
-        [testProject, tasks] << scenarios
+        testProject = SMALL_JAVA_MULTI_PROJECT
```
Comment (wolfs) [inline subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/buildcache/TaskOutputCachingJavaPerformanceTest.groovy]: I guess you didn't want to commit these changes.

Comment (lptr) [reply subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/buildcache/TaskOutputCachingJavaPerformanceTest.groovy]: Good catch, reverted.

**Comment by lptr** [comment]
One issue with this is how to decide if there was a significant improvement. [This got reported](https://builds.gradle.org/viewLog.html?buildId=9642249&buildTypeId=Gradle_Check_IndividualPerformanceScenarioWorkersLinux&tab=buildLog#_state=58,1999,2001), though it should probably not have been:

```text
AWESOME! we're faster than 4.5-20171117235935+0000 :D
Difference: 69 ms faster (69 ms), -0.99%, max regression: 17.007 ms
  Current Gradle median: 6.898 s min: 6.797 s, max: 6.948 s, se: 34.38 ms, sem: 6.277 ms
  > [6.914 s, 6.897 s, 6.919 s, 6.896 s, 6.918 s, 6.802 s, 6.93 s, 6.877 s, 6.903 s, 6.948 s, 6.901 s, 6.87 s, 6.909 s, 6.9 s, 6.875 s, 6.797 s, 6.931 s, 6.912 s, 6.906 s, 6.897 s, 6.926 s, 6.876 s, 6.896 s, 6.883 s, 6.896 s, 6.813 s, 6.889 s, 6.908 s, 6.903 s, 6.895 s]
  Gradle 4.5-20171117235935+0000 median: 6.968 s min: 6.937 s, max: 7.092 s, se: 27.72 ms, sem: 5.061 ms
  > [6.956 s, 6.977 s, 6.971 s, 6.943 s, 6.95 s, 6.979 s, 6.96 s, 6.986 s, 6.963 s, 7.092 s, 6.976 s, 6.99 s, 6.955 s, 6.937 s, 6.975 s, 6.968 s, 6.984 s, 6.974 s, 6.967 s, 6.993 s, 6.962 s, 6.962 s, 6.988 s, 6.962 s, 6.942 s, 6.989 s, 6.945 s, 6.955 s, 6.953 s, 7.003 s]
```

**Comment by lptr** [comment]
This is how reports look like:

image

**Comment by oehme** [comment]
I like the idea! We can probably iron out the threshold based on what we see. Maybe start with "more than one standard error faster". That's much more than we use for regressions, but for regressions we also have the retry to reduce flakiness. 

The results from `master` should probably go to the `#performance` channel.

**Comment by lptr** [comment]
"more than one standard error faster" produces a lot of noise unfortunately. I think 4x standard error could be a good threshold. If it's still too noise, we can reevaluate in the light of evidence.

The other question is who to notify. I'd say `master` and `release` should go to the configured channel, other branches should be ignored. (You'd need to start your own performance build for such a branch, and that means you are probably watching the results. If not, you'll still be notified after you merge.)

**Comment by lptr** [comment]
Performance test build: https://builds.gradle.org/viewQueued.html?itemId=9765371

**Pull request merged**